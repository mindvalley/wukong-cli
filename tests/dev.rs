mod common;

#[test]
fn test_wukong_dev_help() {
    let cmd = common::wukong_raw_command()
        .arg("dev")
        .arg("help")
        .assert()
        .success();

    let output = cmd.get_output();

    insta::assert_snapshot!(std::str::from_utf8(&output.stdout).unwrap());
}

// FIXME: the test is using temporary directory to create config files, but the
//       temporary directory name will be different for each test run, so the
//       snapshot will be different for each test run.
// #[test]
// fn test_wukong_dev_config_lint_success() {
//     let temp = assert_fs::TempDir::new().unwrap();
//     let main_config_file = temp.child("config/config.exs");
//     main_config_file.touch().unwrap();
//
//     main_config_file
//         .write_str(
//             r#"
// use Mix.Config
//
// System.get_env("API_KEY")
// System.fetch_env("API_SECRET")
// System.fetch_env!("API_TOKEN")
//
// # invalid
// import_config "config/dev.exs"
//
// # valid
// if File.exists?("config/dev.exs") do
//   import_config "config/dev.exs"
// end
//
// # invalid
// if File.exists?("config/a.exs") do
//   import_config "config/b.exs"
// end
//
// # valid
// File.exists?("config/dev.exs") && import_config "config/dev.exs"
// # invalid
// File.exists?("config/a.exs") && import_config "config/b.exs"
//
// test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")
//
// # Use Jason for JSON parsing in Phoenix
// config :phoenix, :json_library, Jason
//     "#,
//         )
//         .unwrap();
//
//     let dev_config_file = temp.child("config/dev.exs");
//     dev_config_file.touch().unwrap();
//
//     dev_config_file
//         .write_str(
//             r#"
// use Mix.Config
//
// System.get_env("API_KEY")
// System.fetch_env("API_SECRET")
// System.fetch_env!("API_TOKEN")
//
// # invalid
// import_config "config/dev.exs"
//
// # valid
// if File.exists?("config/dev.exs") do
//   import_config "config/dev.exs"
// end
//
// # invalid
// if File.exists?("config/a.exs") do
//   import_config "config/b.exs"
// end
//
// # valid
// File.exists?("config/dev.exs") && import_config "config/dev.exs"
// # invalid
// File.exists?("config/a.exs") && import_config "config/b.exs"
//
// test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")
//
// # Use Jason for JSON parsing in Phoenix
// config :phoenix, :json_library, Jason
//     "#,
//         )
//         .unwrap();
//
//     let cmd = common::wukong_raw_command()
//         .arg("dev")
//         .arg("config-lint")
//         .arg(temp.path().to_str().unwrap())
//         .assert()
//         .success();
//
//     let output = cmd.get_output();
//
//     insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());
//
//     temp.close().unwrap();
// }

// FIXME: the test is using temporary directory to create elixir project, but the
//       temporary directory name will be different for each test run, so the
//       snapshot will be different for each test run.
// #[test]
// fn test_wukong_dev_config_synthesizer_success() {
//     let server = MockServer::start();
//
//     let secret_api_resp = r#"
// {
//   "data": {
//     "data": {
//       "a.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5",
//       "b.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5",
//       "c.secret.exs": "use Mix.Config\n\nconfig :application, Application.Repo,\n  adapter: Ecto.Adapters.Postgres,\n  username: System.get_env(\"DB_USER\"),\n  password: System.get_env(\"DB_PASS\"),\n  database: \"application_dev\",\n  hostname: \"localhost\",\n  pool_size: 100,\n  queue_target: 5"
//     },
//     "metadata": {
//       "created_time": "2015-02-22T02:24:06.945319214Z",
//       "custom_metadata": {
//         "owner": "xxx",
//         "mission_critical": "false"
//       },
//       "deletion_time": "",
//       "destroyed": false,
//       "version": 2
//     }
//   }
// }"#;
//
//     let verify_token_api_resp = r#"
// {
//   "data": {
//     "expire_time": "2019-12-10T10:10:10.000000Z",
//     "issue_time": "2019-10-10T10:10:10.000000Z"
//     }
// }"#;
//
//     let verify_token_mock = server.mock(|when, then| {
//         when.method(GET)
//             .path_contains(VaultClient::VERIFY_TOKEN)
//             .header("X-Vault-Token", "valid_vault_api_token");
//         then.status(200)
//             .header("content-type", "application/json; charset=UTF-8")
//             .body(verify_token_api_resp);
//     });
//
//     let secret_data_mock = server.mock(|when, then| {
//         when.method(GET).path_contains(VaultClient::FETCH_SECRETS);
//         then.status(200)
//             .header("content-type", "application/json; charset=UTF-8")
//             .body(secret_api_resp);
//     });
//
//     let wk_temp = assert_fs::TempDir::new().unwrap();
//     let wk_config_file = wk_temp.child("config.toml");
//     wk_config_file.touch().unwrap();
//
//     wk_config_file
//         .write_str(
//             format!(
//                 r#"
// [core]
// application = "valid-application"
// wukong_api_url = "{}"
// okta_client_id = "valid-okta-client-id"
//
// [vault]
// api_token = "valid_vault_api_token"
//
// [auth]
// account = "test@email.com"
// subject = "subject"
// id_token = "id_token"
// access_token = "access_token"
// expiry_time = "{}"
// refresh_token = "refresh_token"
//     "#,
//                 server.base_url(),
//                 2.days().from_now().to_rfc3339()
//             )
//             .as_str(),
//         )
//         .unwrap();
//
//     let elixir_temp = assert_fs::TempDir::new().unwrap();
//     let dev_config_file = elixir_temp.child("config/dev.exs");
//     dev_config_file.touch().unwrap();
//
//     dev_config_file
//         .write_str(
//             r#"
// use Mix.Config
//
// System.get_env("API_KEY")
// System.fetch_env("API_SECRET")
// System.fetch_env!("API_TOKEN")
//
// # wukong.mindvalley.dev/config-secrets-location: vault:mv/tech/app/dev#a.secret.exs
// import_config "config/a.secret.exs"
//
// # wukong.mindvalley.dev/config-secrets-location: vault:mv/tech/app/dev#b.secret.exs
// if File.exists?("config/b.secret.exs") do
//   import_config "config/b.secret.exs"
// end
//
// # wukong.mindvalley.dev/config-secrets-location: vault:mv/tech/app/dev#c.secret.exs
// File.exists?("config/c.secret.exs") && import_config "config/c.secret.exs"
//
// test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")
//
// # Use Jason for JSON parsing in Phoenix
// config :phoenix, :json_library, Jason
//     "#,
//         )
//         .unwrap();
//
//     let cmd = common::wukong_raw_command()
//         .arg("dev")
//         .arg("config-synthesizer")
//         .arg(elixir_temp.path().to_str().unwrap())
//         .env("WUKONG_DEV_CONFIG_FILE", wk_config_file.path())
//         .env("WUKONG_DEV_VAULT_API_URL", server.base_url())
//         .assert()
//         .success();
//
//     let output = cmd.get_output();
//
//     insta::assert_snapshot!(std::str::from_utf8(&output.stderr).unwrap());
//
//     verify_token_mock.assert();
//     secret_data_mock.assert();
//
//     wk_temp.close().unwrap();
//     elixir_temp.close().unwrap();
// }
