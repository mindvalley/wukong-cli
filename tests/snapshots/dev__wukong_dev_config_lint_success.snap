---
source: tests/dev.rs
expression: "std::str::from_utf8(&output.stderr).unwrap()"
---

  [31mx[0m no_env_in_dev_config
   ,-[[36;1;4m[TEMP_DIR]/config/dev.exs[0m:3:1]
 [2m3[0m | 
 [2m4[0m | System.get_env("API_KEY")
   : [35;1m^^^^^^^^^^^^|^^^^^^^^^^^^[0m
   :             [35;1m`-- [35;1mDev config must not contains any environment variables[0m[0m
 [2m5[0m | System.fetch_env("API_SECRET")
   `----
[36m  help: [0mUse a static value instead of reading from environment variable. Also if this is a secret, move it to the `dev.secrets.exs` instead.


  [31mx[0m no_env_in_dev_config
   ,-[[36;1;4m[TEMP_DIR]/config/dev.exs[0m:4:1]
 [2m4[0m | System.get_env("API_KEY")
 [2m5[0m | System.fetch_env("API_SECRET")
   : [35;1m^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^[0m
   :                [35;1m`-- [35;1mDev config must not contains any environment variables[0m[0m
 [2m6[0m | System.fetch_env!("API_TOKEN")
   `----
[36m  help: [0mUse a static value instead of reading from environment variable. Also if this is a secret, move it to the `dev.secrets.exs` instead.


  [31mx[0m no_env_in_dev_config
   ,-[[36;1;4m[TEMP_DIR]/config/dev.exs[0m:5:1]
 [2m5[0m | System.fetch_env("API_SECRET")
 [2m6[0m | System.fetch_env!("API_TOKEN")
   : [35;1m^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^[0m
   :                [35;1m`-- [35;1mDev config must not contains any environment variables[0m[0m
 [2m7[0m | 
   `----
[36m  help: [0mUse a static value instead of reading from environment variable. Also if this is a secret, move it to the `dev.secrets.exs` instead.


  [31mx[0m no_env_in_dev_config
    ,-[[36;1;4m[TEMP_DIR]/config/dev.exs[0m:25:1]
 [2m25[0m | 
 [2m26[0m | test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")
    : [35;1m              ^^^^^^^^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^^^^^^^^[0m
    :                                     [35;1m`-- [35;1mDev config must not contains any environment variables[0m[0m
 [2m27[0m | 
    `----
[36m  help: [0mUse a static value instead of reading from environment variable. Also if this is a secret, move it to the `dev.secrets.exs` instead.


  [31mx[0m use_import_config_with_file_exists_checking
    ,-[[36;1;4m[TEMP_DIR]/config/dev.exs[0m:8:1]
 [2m 8[0m | # invalid
 [2m 9[0m | import_config "config/dev.exs"
    : [35;1m^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^[0m
    :                [35;1m`-- [35;1mConsidering checking the existence of the file before importing using `File.exists?/2` function.[0m[0m
 [2m10[0m | 
    `----
[36m  help: [0mUse a static value instead of reading from environment variable. Also if this is a secret, move it to the `dev.secrets.exs` instead.


  [31mx[0m use_import_config_with_file_exists_checking
    ,-[[36;1;4m[TEMP_DIR]/config/dev.exs[0m:16:1]
 [2m16[0m |     # invalid
 [2m17[0m | [35;1m,[0m[35;1m-[0m[35;1m>[0m if File.exists?("config/a.exs") do
 [2m18[0m | [35;1m|[0m     import_config "config/b.exs"
 [2m19[0m | [35;1m|[0m[35;1m-[0m[35;1m>[0m end
    : [35;1m`[0m[35;1m---[0m[35;1m-[0m [35;1mConsidering checking the existence of the file before importing using `File.exists?/2` function.[0m
 [2m20[0m |     
    `----
[36m  help: [0mUse a static value instead of reading from environment variable. Also if this is a secret, move it to the `dev.secrets.exs` instead.


  [31mx[0m use_import_config_with_file_exists_checking
    ,-[[36;1;4m[TEMP_DIR]/config/dev.exs[0m:23:1]
 [2m23[0m | # invalid
 [2m24[0m | File.exists?("config/a.exs") && import_config "config/b.exs"
    : [35;1m^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^^^^^^^^^^^^^^^^[0m
    :                               [35;1m`-- [35;1mConsidering checking the existence of the file before importing using `File.exists?/2` function.[0m[0m
 [2m25[0m | 
    `----
[36m  help: [0mUse a static value instead of reading from environment variable. Also if this is a secret, move it to the `dev.secrets.exs` instead.


  [31mx[0m no_env_in_main_config
   ,-[[36;1;4m[TEMP_DIR]/config/config.exs[0m:3:1]
 [2m3[0m | 
 [2m4[0m | System.get_env("API_KEY")
   : [35;1m^^^^^^^^^^^^|^^^^^^^^^^^^[0m
   :             [35;1m`-- [35;1mMain config must not contains any environment variables[0m[0m
 [2m5[0m | System.fetch_env("API_SECRET")
   `----
[36m  help: [0mMove this config value into either `runtime.exs` if youâ€™re reading them during runtime, or providing a static value.


  [31mx[0m no_env_in_main_config
   ,-[[36;1;4m[TEMP_DIR]/config/config.exs[0m:4:1]
 [2m4[0m | System.get_env("API_KEY")
 [2m5[0m | System.fetch_env("API_SECRET")
   : [35;1m^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^[0m
   :                [35;1m`-- [35;1mMain config must not contains any environment variables[0m[0m
 [2m6[0m | System.fetch_env!("API_TOKEN")
   `----
[36m  help: [0mMove this config value into either `runtime.exs` if youâ€™re reading them during runtime, or providing a static value.


  [31mx[0m no_env_in_main_config
   ,-[[36;1;4m[TEMP_DIR]/config/config.exs[0m:5:1]
 [2m5[0m | System.fetch_env("API_SECRET")
 [2m6[0m | System.fetch_env!("API_TOKEN")
   : [35;1m^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^[0m
   :                [35;1m`-- [35;1mMain config must not contains any environment variables[0m[0m
 [2m7[0m | 
   `----
[36m  help: [0mMove this config value into either `runtime.exs` if youâ€™re reading them during runtime, or providing a static value.


  [31mx[0m no_env_in_main_config
    ,-[[36;1;4m[TEMP_DIR]/config/config.exs[0m:25:1]
 [2m25[0m | 
 [2m26[0m | test_domain = System.get_env("TEST_DOMAIN", "mv.test.com")
    : [35;1m              ^^^^^^^^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^^^^^^^^[0m
    :                                     [35;1m`-- [35;1mMain config must not contains any environment variables[0m[0m
 [2m27[0m | 
    `----
[36m  help: [0mMove this config value into either `runtime.exs` if youâ€™re reading them during runtime, or providing a static value.

Total time taken: [DURATION_IN_MS] ([DURATION_IN_MS] to load, [DURATION_IN_MS] running 3 checks)
Total files: 2

